src/query_plan/fetch_dependency_graph_processor.rs

```diff
@@ -1,2 +1,398 @@
+/// since this affect overall latency directly.
+/// The exact number is a tad  arbitrary however.
+const PIPELINING_COST: QueryPlanCost = 100;
+
+pub(crate) struct FetchDependencyGraphToQueryPlanProcessor {
+    config: Arc<QueryPlannerConfig>,
+    variable_definitions: Vec<Node<VariableDefinition>>,
+    fragments: Option<RebasedFragments>,
+    operation_name: Option<Name>,
+    assigned_defer_labels: Option<HashSet<NodeStr>>,
+    counter: u32,
+}
+
+// TODO
+pub(crate) struct RebasedFragments;
+
+/// Computes the cost of a Plan.
+///
+/// A plan is essentially some mix of sequences and parallels of fetches. And the plan cost
+/// is about minimizing both:
+///  1. The expected total latency of executing the plan. Typically, doing 2 fetches in
+///    parallel will most likely have much better latency then executing those exact same
+///    fetches in sequence, and so the cost of the latter must be greater than that of
+///    the former.
+///  2. The underlying use of resources. For instance, if we query 2 fields and we have
+///    the choice between getting those 2 fields from a single subgraph in 1 fetch, or
+///    get each from a different subgraph with 2 fetches in parallel, then we want to
+///    favor the former as just doing a fetch in and of itself has a cost in terms of
+///    resources consumed.
+///
+/// Do note that at the moment, this cost is solely based on the "shape" of the plan and has
+/// to make some conservative assumption regarding concrete runtime behaviour. In particular,
+/// it assumes that:
+///  - all fields have the same cost (all resolvers take the same time).
+///  - that field cost is relative small compare to actually doing a subgraph fetch. That is,
+///    it assumes that the networking and other query processing costs are much higher than
+///    the cost of resolving a single field. Or to put it more concretely, it assumes that
+///    a fetch of 5 fields is probably not too different from than of 2 fields.
 pub(crate) struct FetchDependencyGraphToCostProcessor;
+
+/// Generic interface for "processing" a (reduced) dependency graph of fetch groups
+/// (a `FetchDependencyGraph`).
+///
+/// The processor methods will be called in a way that "respects" the dependency graph.
+/// More precisely, a reduced fetch group dependency graph can be expressed
+/// as an alternance of parallel branches and sequences of groups
+/// (the roots needing to be either parallel or
+/// sequential depending on whether we represent a `query` or a `mutation`),
+/// and the processor will be called on groups in such a way.
+pub(crate) trait FetchGroupProcessor<TProcessed, TDeferred> {
+    fn on_fetch_group(
+        &mut self,
+        group: &mut FetchDependencyGraphNode,
+        handled_conditions: &Conditions,
+    ) -> Result<TProcessed, FederationError>;
+    fn on_conditions(&mut self, conditions: &Conditions, value: TProcessed) -> TProcessed;
+    fn reduce_parallel(&mut self, values: &[TProcessed]) -> TProcessed;
```

Comment:
Something for later, but I think all the callsites of `reduce_parallel()` and `reduce_sequence()` will pass ownership of `TProcessed` (but cloning is cheap for `PlanNode`s, so this doesn't really matter).

Reply:
Changed to take iterators of owned `TProcessed` values.